Refactor cpcode matcher

simpsora opened the pull request:
The old one didn't work at all, it only looked at the `X-Cache-Key`
header, rather than decoding the values within it and matching on those.

This change does several things:

* Splits out all the values from the `X-Cache-Key` header, then validates that the passed-in cpcode is present in the cpcode field in that header.
* Uses a custom failure message to return the actual cpcode found to the caller, using thread local storage.
* Removes the 200 response code requirement from the cpcode matcher, as the test is about cpcodes, not response status.
* Adds more Akamai debug headers (and makes the existing list more readable)
* Logs the network name at the beginning of a test run

/cc @patrobinson @jacobbednarz

simpson-ross committed:
Add more debug headers

Also make the list more readable.

simpson-ross committed:
Refactor the cpcode matcher

The old one didn't work at all, it only looked at the `X-Cache-Key`
header, rather than decoding the values within it.

This change splits out all the values from that header, then validates
that the passed-in cpcode is present in the cpcode field in that header.

It also uses a custom failure message to return the actual cpcode found
to the caller, using thread local storage.

Finally, it removes the 200 response code requirement, as the test is
about cpcodes, not response status.

simpson-ross committed:
Log the network when setting it

So it's clear from the test run which network is in use.

simpson-ross committed:
Bump ruby version to 2.3.0

patrobinson commented on .travis.yml:
```diff
@@ -1,5 +1,3 @@
- - 2.1.2
- - 2.1.8
- - 2.2.4
+ - 2.3.0
```
Removing the other versions may make it harder to get these changes merged upstream

simpsora replied:
Yep.  Since upstream has diverged a fair bit (with that giant refactor PR they had), I don't expect to get these changes merged.  

Some of the code I'm using requires later versions.  I could write it to use older versions, but in the interest of GSD I'm just making it work.

Later on I want to give some thought to what we should do long-term about this project and its divergence from upstream.

patrobinson replied:
Happy to help, I've spent a LOT of time working on this library

patrobinson reviewed (approved):
On green

simpson-ross committed:
Handle missing values when finding cpcode

Some values might be nil.

simpson-ross committed:
Fix newly-broken cpcode specs

Previous cpcode support was very inadequate.  These specs have been
fixed to reflect the new mechanism of parsing the cpcode out of
`X-Cache-Key`.

simpson-ross committed:
Move gem deps into gemspec

Rather than defining them in `Gemfile`.
